File: app/src/main/java/dev/leonlatsch/olivia/database/interfaces/ContactInterface.java
package dev.leonlatsch.olivia.database.interfaces;

import com.activeandroid.query.Delete;
import com.activeandroid.query.Select;
import com.activeandroid.query.Update;

import java.util.List;

import dev.leonlatsch.olivia.database.model.Contact;
import dev.leonlatsch.olivia.database.model.User;
import dev.leonlatsch.olivia.rest.dto.UserDTO;

public class ContactInterface extends BaseInterface {

    private static final String QUEUE_UID_WHERE = "uid = ?";

    private static ContactInterface contactInterface; // Singleton


    private ContactInterface() {}

    public Contact getContact(String uid) {
        return new Select().from(Contact.class).where(QUEUE_UID_WHERE, uid).executeSingle();
    }

    public List<Contact> getALl() {
        return new Select().from(Contact.class).execute();
    }

    public void delete(Contact contact) {
        contact.delete();
    }

    public void deleteAll() {
        new Delete().from(Contact.class).execute();
    }

    public String save(UserDTO userDTO, String publicKey) {
        return save(getDatabaseMapper().toModel(userDTO), publicKey);
    }

    public String save(User user, String publicKey) {
        Contact contact = getDatabaseMapper().toContact(user);
        contact.setPublicKey(publicKey);
        contact.setUid(user.getUid());
        return save(contact);
    }

    public String save(Contact contact) {
        Contact saved = new Select().from(Contact.class).where(QUEUE_UID_WHERE, contact.getUid()).executeSingle();

        if (saved != null) {
            saved.delete();
        }

        contact.save();
        return contact.getUid();
    }

    public void updateContact(Contact contact) {
        if (contact.getProfilePicTn() != null) {
            new Update(Contact.class).set("uid = ?, username = ?, profile_pic_tn = ?, public_key = ?",
                    contact.getUid(), contact.getUsername(), contact.getProfilePicTn(), contact.getPublicKey())
                    .where(QUEUE_UID_WHERE, contact.getUid()).execute();
        } else {
            new Update(Contact.class).set("uid = ?, username = ?, public_key = ?",
                    contact.getUid(), contact.getUsername(), contact.getPublicKey())
                    .where(QUEUE_UID_WHERE, contact.getUid()).execute();
        }
    }

    public static ContactInterface getInstance() {
        if (contactInterface == null) {
            contactInterface = new ContactInterface();
        }

        return contactInterface;
    }
}

File: app/src/main/java/dev/leonlatsch/kolibri/database/interfaces/ContactInterface.kt
package dev.leonlatsch.kolibri.database.interfaces

import com.activeandroid.query.Delete
import com.activeandroid.query.Select
import com.activeandroid.query.Update
import dev.leonlatsch.kolibri.database.DatabaseMapper

import dev.leonlatsch.kolibri.database.model.Contact
import dev.leonlatsch.kolibri.database.model.User
import dev.leonlatsch.kolibri.rest.dto.UserDTO

/**
 * Database interface to persist contacts
 *
 * @author Leon Latsch
 * @since 1.0.0
 */
object ContactInterface {

    private const val QUEUE_UID_WHERE = "uid = ?"

    val aLl: List<Contact>
        get() = Select().from(Contact::class.java).execute()

    fun getContact(uid: String): Contact {
        return Select().from(Contact::class.java).where(QUEUE_UID_WHERE, uid).executeSingle()
    }

    fun delete(contact: Contact) {
        contact.delete()
    }

    fun delete(uid: String) {
        Delete().from(Contact::class.java).where(QUEUE_UID_WHERE, uid).execute<Contact>()
    }

    fun deleteAll() {
        Delete().from(Contact::class.java).execute<Contact>()
    }

    fun save(userDTO: UserDTO, publicKey: String): String? {
        return save(DatabaseMapper.toModel(userDTO), publicKey)
    }

    fun save(user: User?, publicKey: String): String? {
        val contact = DatabaseMapper.toContact(user)
        contact?.publicKey = publicKey
        contact?.uid = user?.uid
        return save(contact)
    }

    fun save(contact: Contact?): String? {
        val saved = Select().from(Contact::class.java).where(QUEUE_UID_WHERE, contact?.uid).executeSingle<Contact>()

        if (saved != null) {
            saved!!.delete()
        }

        contact?.save()
        return contact?.uid
    }

    fun updateContact(contact: Contact) {
        if (contact.profilePicTn != null) {
            Update(Contact::class.java).set("uid = ?, username = ?, profile_pic_tn = ?, public_key = ?",
                    contact.uid, contact.username, contact.profilePicTn, contact.publicKey)
                    .where(QUEUE_UID_WHERE, contact.uid).execute()
        } else {
            Update(Contact::class.java).set("uid = ?, username = ?, public_key = ?",
                    contact.uid, contact.username, contact.publicKey)
                    .where(QUEUE_UID_WHERE, contact.uid).execute()
        }
    }
}
